composition-js/src/merging/merge.ts

```diff
@@ -2993,22 +2989,24 @@ class Merger {
     if (dest instanceof InterfaceType) {
       // we need to propagate access control from implementing types upward to the interface
       for (const { name, nameInSupergraph } of this.accessControlDirectivesInSupergraph) {
-        if (nameInSupergraph) {
-          let additionalSources = this.accessControlAdditionalSources().get(`${dest.coordinate}_${name}`);
-          if (!additionalSources) {
-            additionalSources = [];
-          }
+        let additionalSources = this.accessControlAdditionalSources().get(`${dest.coordinate}_${name}`);
+        if (!additionalSources) {
+          additionalSources = [];
+        }
+        // WARNING: In order to propagate access control directive requirements, we are hijacking existing merge
+        // directive logic by providing "additional sources" from their interfaces/implementations. This means
+        // that their corresponding subgraphIndex will be incorrect and shouldn't be used/relied on when merging
+        // access control directives.
+        if (names.has(nameInSupergraph) || additionalSources.length) {
```

Comment:
`additionalSources.length > 0` is clearer.

Reply:
[878d2b6](https://github.com/apollographql/federation/pull/3337/commits/878d2b6dc7bdf9d4c7a00d5feed935c0d7edd8b9)